File: backend/app/analytics/router.py
"""
Analytics Module — Router
"""
import uuid
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query

from app.analytics.service import AnalyticsService, get_analytics_service

router = APIRouter()


@router.get("/property/{property_id}")
async def get_property_analytics(
    property_id: str,
    year: int = Query(..., ge=2020, le=2100),
    service: AnalyticsService = Depends(get_analytics_service),
):
    """Get comprehensive property performance analytics (short-TTL cached)."""
    from app.ai.cache import analytics_cache, data_fingerprint

    fp = await data_fingerprint(service.session)
    key = f"prop:{property_id}:{year}:{fp}"
    cached = analytics_cache.get(key)
    if cached is not None:
        return cached
    result = await service.get_property_analytics(uuid.UUID(property_id), year)
    analytics_cache.set(key, result)
    return result


@router.get("/portfolio")
async def get_portfolio_analytics(
    year: int | None = Query(None, ge=2020, le=2100),
    start_date: date | None = Query(None),
    end_date: date | None = Query(None),
    service: AnalyticsService = Depends(get_analytics_service),
):
    """Get portfolio-wide analytics (all properties).

    Either `year` (defaults to the current year) or a `start_date`/`end_date`
    pair selects the period.
    """
    if (start_date is None) != (end_date is None):
        raise HTTPException(
            status_code=422,
            detail="start_date and end_date must be provided together",
        )
    if start_date and end_date and start_date > end_date:
        raise HTTPException(
            status_code=422,
            detail="start_date must not be after end_date",
        )
    from app.ai.cache import analytics_cache, data_fingerprint

    fp = await data_fingerprint(service.session)
    key = f"portfolio:{year}:{start_date}:{end_date}:{fp}"
    cached = analytics_cache.get(key)
    if cached is not None:
        return cached
    result = await service.get_portfolio_analytics(
        year=year, start_date=start_date, end_date=end_date
    )
    analytics_cache.set(key, result)
    return result


@router.get("/property/{property_id}/health")
async def get_property_health(
    property_id: str,
    service: AnalyticsService = Depends(get_analytics_service),
):
    """Get property health score (0-100)."""
    return await service.get_property_health_score(
        uuid.UUID(property_id)
    )

File: backend-rs/src/analytics/router.rs
//! Analytics HTTP routes, mirroring `backend/app/analytics/router.py`.
//!
//! GET /api/v1/analytics/property/{id}?year=YYYY
//! GET /api/v1/analytics/property/{id}/health
//! GET /api/v1/analytics/portfolio?year=YYYY | start_date&end_date

use axum::extract::{Path, Query, State};
use axum::routing::get;
use axum::{Json, Router};
use serde::Deserialize;
use serde_json::Value;
use uuid::Uuid;

use crate::analytics::service;
use crate::core::error::AppError;
use crate::core::state::AppState;

#[derive(Debug, Deserialize)]
struct PropertyQuery {
    year: i32,
}

#[derive(Debug, Default, Deserialize)]
struct PortfolioQuery {
    year: Option<i32>,
    start_date: Option<String>,
    end_date: Option<String>,
}

pub fn build_router() -> Router<AppState> {
    Router::new()
        .route("/property/{id}", get(property_analytics))
        .route("/property/{id}/health", get(property_health))
        .route("/portfolio", get(portfolio_analytics))
}

async fn property_analytics(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Query(q): Query<PropertyQuery>,
) -> Result<Json<Value>, AppError> {
    let property_id = id.to_string();
    let fp = service::data_fingerprint(&state.pool).await?;
    let key = format!("prop:{property_id}:{}:{fp}", q.year);
    if let Some(cached) = service::cache_get(&key) {
        return Ok(Json(cached));
    }
    let result = service::get_property_analytics(&state.pool, &property_id, q.year).await?;
    service::cache_set(&key, result.clone());
    Ok(Json(result))
}

async fn property_health(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<Value>, AppError> {
    let result = service::get_property_health_score(&state.pool, &id.to_string()).await?;
    Ok(Json(result))
}

async fn portfolio_analytics(
    State(state): State<AppState>,
    Query(q): Query<PortfolioQuery>,
) -> Result<Json<Value>, AppError> {
    if (q.start_date.is_some()) != (q.end_date.is_some()) {
        return Err(AppError::Validation(
            "start_date and end_date must be provided together".into(),
        ));
    }
    if let (Some(s), Some(e)) = (q.start_date.as_deref(), q.end_date.as_deref()) {
        if s > e {
            return Err(AppError::Validation(
                "start_date must not be after end_date".into(),
            ));
        }
    }

    let fp = service::data_fingerprint(&state.pool).await?;
    let key = format!(
        "portfolio:{:?}:{:?}:{:?}:{fp}",
        q.year, q.start_date, q.end_date
    );
    if let Some(cached) = service::cache_get(&key) {
        return Ok(Json(cached));
    }
    let result = service::get_portfolio_analytics(
        &state.pool,
        q.year,
        q.start_date.as_deref(),
        q.end_date.as_deref(),
    )
    .await?;
    service::cache_set(&key, result.clone());
    Ok(Json(result))
}
